src/main/java/com/google/devtools/build/lib/bazel/repository/LocalConfigPlatformFunction.java

```diff
@@ -0,0 +1,160 @@
+import com.google.devtools.build.lib.analysis.RuleDefinition;
+import com.google.devtools.build.lib.packages.Rule;
+import com.google.devtools.build.lib.rules.repository.RepositoryDirectoryValue;
+import com.google.devtools.build.lib.rules.repository.RepositoryDirectoryValue.Builder;
+import com.google.devtools.build.lib.rules.repository.RepositoryFunction;
+import com.google.devtools.build.lib.util.CPU;
+import com.google.devtools.build.lib.util.OS;
+import com.google.devtools.build.lib.vfs.Path;
+import com.google.devtools.build.skyframe.SkyFunction.Environment;
+import com.google.devtools.build.skyframe.SkyFunctionException.Transience;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/** Create a local repository that describes the auto-detected host platform. */
+public class LocalConfigPlatformFunction extends RepositoryFunction {
+
+  @Override
+  public boolean isLocal(Rule rule) {
+    return true;
+  }
+
+  @Override
+  public Class<? extends RuleDefinition> getRuleDefinition() {
+    return LocalConfigPlatformRule.class;
+  }
+
+  @Override
+  public Builder fetch(
+      Rule rule,
+      Path outputDirectory,
+      BlazeDirectories directories,
+      Environment env,
+      Map<String, String> markerData)
+      throws RepositoryFunctionException {
+
+    CPU hostCpu = CPU.getCurrent();
+    OS hostOs = OS.getCurrent();
+
+    List<String> constraints = generateConstraints(hostCpu, hostOs);
+
+    try {
+      outputDirectory.createDirectoryAndParents();
+      RepositoryFunction.writeFile(
+          outputDirectory, "WORKSPACE", workspaceFileContent(rule.getName()));
+      RepositoryFunction.writeFile(
+          outputDirectory, "BUILD.bazel", buildFileContent(rule.getName()));
+      RepositoryFunction.writeFile(
+          outputDirectory, "constraints.bzl", constraintFileContent(constraints));
+    } catch (IOException e) {
+      throw new RepositoryFunctionException(
+          new IOException("Could not create content for " + rule.getName() + ": " + e.getMessage()),
+          Transience.TRANSIENT);
+    }
+
+    // Return the needed info.
+    return RepositoryDirectoryValue.builder().setPath(outputDirectory);
+  }
+
+  private static List<String> generateConstraints(CPU hostCpu, OS hostOs) {
+    return ImmutableList.of(cpuToConstraint(hostCpu), osToConstraint(hostOs));
+  }
+
+  @Nullable
+  private static String cpuToConstraint(CPU cpu) {
+    switch (cpu) {
+      case X86_32:
+        return "@bazel_tools//platforms:x86_32";
+      case X86_64:
+        return "@bazel_tools//platforms:x86_64";
+      case PPC:
+        return "@bazel_tools//platforms:ppc";
+      case ARM:
+        return "@bazel_tools//platforms:arm";
+      case AARCH64:
+        return "@bazel_tools//platforms:aarch64";
+      case S390X:
+        return "@bazel_tools//platforms:s390x";
+    }
+    // Unknown, so skip it.
+    return null;
```

Comment:
Won't this crash when fed into ImmutableList? Is there a cleaner error mode?

Reply:
There's a null check where it is used (in constraintFileContent).

It's tricky to figure out how to signal not to use anything. I could switch to using Optional, or pass in the List and optionally add an item. Thoughts?

Reply:
Doesn't ImmutableList.of() crash if one of its inputs is null?

Reply:
You're right! Fixed.